src/Symfony/Component/Routing/Loader/YamlFileLoader.php

```diff
@@ -184,40 +214,50 @@ protected function parseImport(RouteCollection $collection, array $config, $path
      * @param string $name   The config key
      * @param string $path   The loaded file path
      *
-     * @throws \InvalidArgumentException If one of the provided config keys is not supported,
-     *                                   something is missing or the combination is nonsense
+     * @throws InvalidArgumentException If one of the provided config keys is not supported,
+     *                                  something is missing or the combination is nonsense
      */
     protected function validate($config, $name, $path)
     {
         if (!is_array($config)) {
-            throw new \InvalidArgumentException(sprintf('The definition of "%s" in "%s" must be a YAML array.', $name, $path));
+            throw new InvalidArgumentException(sprintf('The definition of "%s" in "%s" must be a YAML array.', $name, $path));
         }
         if ($extraKeys = array_diff(array_keys($config), self::$availableKeys)) {
-            throw new \InvalidArgumentException(sprintf(
+            throw new InvalidArgumentException(sprintf(
                 'The routing file "%s" contains unsupported keys for "%s": "%s". Expected one of: "%s".',
                 $path, $name, implode('", "', $extraKeys), implode('", "', self::$availableKeys)
             ));
         }
-        if (isset($config['resource']) && isset($config['path'])) {
-            throw new \InvalidArgumentException(sprintf(
-                'The routing file "%s" must not specify both the "resource" key and the "path" key for "%s". Choose between an import and a route definition.',
+        if (isset($config['resource']) && (isset($config['path']) || isset($config['locales']))) {
```

Comment:
I would use 2 different `if`, to be able to give the right error message for each case